datastore/api/snippets.py

```diff
@@ -0,0 +1,820 @@
+    query = client.query(kind='Task')
+    query.add_filter(
+        'created', '>', datetime.datetime(1990, 1, 1))
+    query.add_filter(
+        'created', '<', datetime.datetime(2000, 12, 31, 23, 59, 59))
+    tasks = list(query.fetch())
+    # [END inequality_range]
+
+    return tasks
+
+
+def inequality_invalid(client):
+    try:
+        # [START inequality_invalid]
+        query = client.query(kind='Task')
+        query.add_filter(
+            'created', '>', datetime.datetime(1990, 1, 1))
+        query.add_filter(
+            'priority', '>', 3)
+        tasks = list(query.fetch())
+        # [END inequality_invalid]
+
+        return tasks
+
+    except gcloud.exceptions.BadRequest:
+        pass
+
+
+def equal_and_inequality_range(client):
+    # [START equal_and_inequality_range]
+    query = client.query(kind='Task')
+    query.add_filter('priority', '=', 4)
+    query.add_filter('done', '=', False)
+    query.add_filter(
+        'created', '>', datetime.datetime(1990, 1, 1))
+    query.add_filter(
+        'created', '<', datetime.datetime(2000, 12, 31, 23, 59, 59))
+    tasks = list(query.fetch())
+    # [END equal_and_inequality_range]
+
+    return tasks
+
+
+def inequality_sort(client):
+    # [START inequality_sort]
+    query = client.query(kind='Task')
+    query.add_filter('priority', '>', 3)
+    query.order = ['priority', 'created']
+    tasks = list(query.fetch())
+    # [END inequality_sort]
+
+    return tasks
+
+
+def inequality_sort_invalid_not_same(client):
+    try:
+        # [START inequality_sort_invalid_not_same]
+        query = client.query(kind='Task')
+        query.add_filter('priority', '>', 3)
+        query.order = ['created']
+        tasks = list(query.fetch())
+        # [END inequality_sort_invalid_not_same]
+
+        return tasks
+
+    except gcloud.exceptions.BadRequest:
+        pass
+
+
+def inequality_sort_invalid_not_first(client):
+    try:
+        # [START inequality_sort_invalid_not_first]
+        query = client.query(kind='Task')
+        query.add_filter('priority', '>', 3)
+        query.order = ['created', 'priority']
+        tasks = list(query.fetch())
+        # [END inequality_sort_invalid_not_first]
+
+        return tasks
+
+    except gcloud.exceptions.BadRequest:
+        pass
+
+
+def array_value_inequality_range(client):
+    # [START array_value_inequality_range]
+    query = client.query(kind='Task')
+    query.add_filter('tag', '>', 'learn')
+    query.add_filter('tag', '<', 'math')
+    tasks = list(query.fetch())
+    # [END array_value_inequality_range]
+
+    return tasks
+
+
+def array_value_equality(client):
+    # [START array_value_equality]
+    query = client.query(kind='Task')
+    query.add_filter('tag', '=', 'fun')
+    query.add_filter('tag', '=', 'programming')
+    tasks = list(query.fetch())
+    # [END array_value_equality]
+
+    return tasks
+
+
+def exploding_properties(client):
+    # [START exploding_queries]
```

Comment:
The documentation explains it (e.g., these snippets are not intended to be consumed without the documentation). Essentially having a indexed on two properties that are arrays leads to the index containing the combination of all elements of the two arrays.